server/src/main/java/org/elasticsearch/index/codec/tsdb/es819/ES819TSDBDocValuesConsumer.java

```diff
@@ -138,89 +143,125 @@ private long[] writeField(FieldInfo field, TsdbDocValuesProducer valuesProducer,
-                int bufferSize = 0;
-                final TSDBDocValuesEncoder encoder = new TSDBDocValuesEncoder(ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE);
-                values = valuesProducer.getSortedNumeric(field);
-                final int bitsPerOrd = maxOrd >= 0 ? PackedInts.bitsRequired(maxOrd - 1) : -1;
-                for (int doc = values.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = values.nextDoc()) {
-                    final int count = values.docValueCount();
-                    for (int i = 0; i < count; ++i) {
-                        buffer[bufferSize++] = values.nextValue();
-                        if (bufferSize == ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE) {
-                            indexWriter.add(data.getFilePointer() - valuesDataOffset);
-                            if (maxOrd >= 0) {
-                                encoder.encodeOrdinals(buffer, data, bitsPerOrd);
-                            } else {
-                                encoder.encode(buffer, data);
+        IndexOutput disiTempOutput = null;
+        String skipListTempFileName = null;
+        IndexedDISIBuilder docIdSetBuilder = null;
+        try {
+            if (numValues > 0) {
+                // Special case for maxOrd of 1, signal -1 that no blocks will be written
+                meta.writeInt(maxOrd != 1 ? ES819TSDBDocValuesFormat.DIRECT_MONOTONIC_BLOCK_SHIFT : -1);
+                final ByteBuffersDataOutput indexOut = new ByteBuffersDataOutput();
+                final DirectMonotonicWriter indexWriter = DirectMonotonicWriter.getInstance(
+                    meta,
+                    new ByteBuffersIndexOutput(indexOut, "temp-dv-index", "temp-dv-index"),
+                    1L + ((numValues - 1) >>> ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SHIFT),
+                    ES819TSDBDocValuesFormat.DIRECT_MONOTONIC_BLOCK_SHIFT
+                );
+
+                final long valuesDataOffset = data.getFilePointer();
+                // Special case for maxOrd of 1, skip writing the blocks
+                if (maxOrd != 1) {
+                    final long[] buffer = new long[ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE];
+                    int bufferSize = 0;
+                    final TSDBDocValuesEncoder encoder = new TSDBDocValuesEncoder(ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE);
+                    values = valuesProducer.getSortedNumeric(field);
+                    final int bitsPerOrd = maxOrd >= 0 ? PackedInts.bitsRequired(maxOrd - 1) : -1;
+                    if (numDocsWithValue != 0 && numDocsWithValue != maxDoc) {
+                        // TODO: which IOContext should be used here?
+                        disiTempOutput = dir.createTempOutput(data.getName(), "disi", IOContext.DEFAULT);
+                        skipListTempFileName = disiTempOutput.getName();
+                        docIdSetBuilder = new IndexedDISIBuilder(disiTempOutput, IndexedDISI.DEFAULT_DENSE_RANK_POWER);
```

Comment:
Forked the write side of IndexedDISI and turn it into a builder. The downside is that we live with another forked piece of code and we only forked the write side. I'm not happy about that.

Initially I collected the the docids in a roaring bitset, but that doesn't scale given that we keep track of doc ids with value in jvm heap. 

Is there a better alternative?